src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -445,14 +445,15 @@ private ValueTask OnEndStreamReceived()
                     return ProcessDataFrameAsync(payload);
                 case Http3FrameType.Headers:
                     return ProcessHeadersFrameAsync(application, payload);
-                // need to be on control stream
-                case Http3FrameType.DuplicatePush:
-                case Http3FrameType.PushPromise:
                 case Http3FrameType.Settings:
-                case Http3FrameType.GoAway:
                 case Http3FrameType.CancelPush:
+                case Http3FrameType.GoAway:
                 case Http3FrameType.MaxPushId:
-                    throw new Http3ConnectionException("HTTP_FRAME_UNEXPECTED");
+                    // These frames need to be on a control stream
+                    throw new Http3StreamErrorException(CoreStrings.FormatHttp3ErrorUnsupportedFrameOnRequestStream(_incomingFrame.FormattedType), Http3ErrorCode.UnexpectedFrame);
```

Comment:
Does this abort the entire connection? Or just the current stream?

Reply:
Reading the spec, I believe these should be connection level errors:
https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-7.2.1
"DATA frames MUST be associated with an HTTP request or response. If a DATA frame is received on a control stream, the recipient MUST respond with a connection error of type H3_FRAME_UNEXPECTED; see Section 8."

Reply:
Changed exceptions thrown. Aborting the overall connection from a stream doesn't look like it exists today. I'd like to add that in a future PR so this one doesn't turn into a monster.